lib/spack/spack/spec.py

```diff
@@ -1923,8 +1932,13 @@ def to_node_dict(self, hash=ht.dag_hash):
             if hasattr(variant, '_patches_in_order_of_appearance'):
                 d['patches'] = variant._patches_in_order_of_appearance
 
-        if self._concrete and hash.package_hash:
-            package_hash = self.package_hash()
+        if self._concrete and hash.package_hash and self._package_hash:
+            # We use the attribute here instead of `self.package_hash()` because this
+            # should *always* be assignhed at concretization time. We don't want to try
+            # to compute a package hash for concrete spec where a) the package might not
+            # exist, or b) the `dag_hash` didn't include the package hash when the spec
+            # was concretized.
+            package_hash = self._package_hash
```

Comment:
This is related to #47973 The fact is, we should ignore the package hash only on "old" concrete specs. This conditional, as written right now, doesn't distinguish a case of an "old" spec, from a case of a "new" spec where the `_package_hash` is missing for whatever reason. The latter case may lead to hash inconsistencies.